Refactoring job dispath

mortalisk opened the pull request:
**Overview:**
- Removed code paths that essentially did the same things (e.g. [`kill_job_and_EXIT`](https://github.com/equinor/libres/blob/5c88920febfb97a18c6a2d4b0e51bfd2de8f0241/bin/job_dispatch.py#L102)/[`JobManager.exit`](https://github.com/equinor/libres/blob/5c88920febfb97a18c6a2d4b0e51bfd2de8f0241/python/res/job_queue/job_manager.py#L470), [`JobManager.execJob`](https://github.com/equinor/libres/blob/5c88920febfb97a18c6a2d4b0e51bfd2de8f0241/python/res/job_queue/job_manager.py#L345)/[`exec_job`](https://github.com/equinor/libres/blob/5c88920febfb97a18c6a2d4b0e51bfd2de8f0241/bin/job_dispatch.py#L78))
- Created a `JobRunner` that has the responsibility of running the individual `Job`s.
- Created a `Job` that implements running a job.
- Created a `Reporter` which is subclassed by `Networking`,
  `File` and `Interactive`.
- Created a class of `Message`s that allows both `JobRunner` and `Job` to communicate with the reporters.
- Simplified testing and added tests for introduced classes.
- Removed job spawning using `os.fork` and `os.execvp`, replaced it with `Popen`.

Resolves #658  
Resolves #661 
Resolves #646 
Resolves #651 
Resolves https://github.com/equinor/libres/issues/316

andreabrambilla commented on python/res/job_queue/job.py:
```diff
@@ -8,7 +8,7 @@
 #
 #  ERT is distributed in the hope that it will be useful, but WITHOUT ANY
-#  WARRANTY; without even the implied warranty of MERCHANTABILITY or
+#  WARRANTY; without even the implied warranty of MEargListRCHANTABILITY or
```
what?

jondequinor committed:
Refactor job_dispatch

- Removed code paths that almost did the same things
- Rewrote the JobRunner/Job API to be run() only
- Introduced the concept of a Reporter, of which there are Networking,
  File and Interactive.
- Reporters subscribe to JobRunner and Job status messages
- Simplified unit testing overall

sondreso committed:
Remove license check for jobs

ManInFez commented on python/job_runner/reporting/__init__.py:
```diff
@@ -0,0 +1,7 @@
+"""
+from .file import File
+from .interactive import Interactive
+from .network import Network
```
no new line on file

ManInFez commented on python/job_runner/reporting/file.py:
```diff
@@ -0,0 +1,184 @@
+    EXIT_file = "EXIT"
+    STATUS_file = "STATUS"
+    OK_file = "OK"
+    STATUS_FILE = "status.json"
```
couldnt jobs.json be defined here, instead of injected as magic strings everywhere?

ManInFez replied:
Also all the is "_file" except status which is "_FILE"

ManInFez replied:
I would maybe call STATUS_FILE STATUS_json, and perhaps fetch the values from their C counterpart in job_queue/forward_model.cpp

jondequinor replied:
> couldnt jobs.json be defined here, instead of injected as magic strings everywhere?

It could, but it is really a `JobRunner` thing.



> … perhaps fetch the values from their C counterpart in job_queue/forward_model.cpp

`#defines` are garbled compile time, so they're not in the DLL (at least not accessible using the original name).

And as Markus discuss in https://github.com/equinor/ert/issues/434#issuecomment-524188452 the real fix to this might be to drop the file system as means for communication between the queue and job_dispatch altogether?

ManInFez commented on python/job_runner/reporting/file.py:
```diff
@@ -0,0 +1,184 @@
+        self.status_dict = {}
+        self.node = socket.gethostname()
+
+    def report(self, msg, sync_disc_timeout=10):
```
File reporter has different signature(... , msg, .. ) Interactive and network ( ... , status, ... )

ManInFez commented on python/job_runner/runner.py:
```diff
@@ -0,0 +1,87 @@
+            yield init_message
+
+        all_jobs_successful = True
+        for job in job_queue:
```
Job runner should terminate if one job is not successful

ManInFez replied:
There should be a test for this ;)

ManInFez replied:
Taken care of by job_dispatch.py

jondequinor replied:
Yes, let's have `job_dispatch` kill on unsuccessful `Finish`, and have file reporter report everything it needs to on `Exited`.

ManInFez commented on python/job_runner/reporting/network.py:
```diff
@@ -0,0 +1,115 @@
+            requests.post(url, timeout=3,
+                          headers={"Content-Type": "application/json"},
+                          data=data, proxies=proxies)
+        except:  # noqa
```
Perhaps report somewhere if post is failing? And shouldn't we now what exception is being thrown?

jondequinor replied:
I agree, but let's do this as its own task: https://github.com/equinor/libres/issues/676

ManInFez commented on python/job_runner/runner.py:
```diff
@@ -0,0 +1,87 @@
+    def __init__(self, jobs_file="jobs.json"):
+        try:
+            with open(jobs_file, "r") as json_file:
+                self.jobs_data = json.load(json_file)
```
self.jobs_data does not have to be object attribute as it is not used anywhere else outside this scope

jondequinor replied:
It is used in the `File` reporter: https://github.com/mortalisk/libres/blob/refactor_job_dispatch/python/job_runner/reporting/file.py#L134

ManInFez replied:
Then there is a typo: jobs_data  != job_data

ManInFez replied:
Correction this is a runner object, not a job object...

ManInFez commented on python/job_runner/runner.py:
```diff
@@ -0,0 +1,87 @@
+        self.ert_pid = self.jobs_data.get("ert_pid")
+        self.global_environment = self.jobs_data.get("global_environment")
+        self.global_update_path = self.jobs_data.get("global_update_path")
+        self.job_data_list = self.jobs_data["jobList"]
```
job_data_list is also not used outside this scope

jondequinor committed:
Simplify exit status, add constants, minor fixes

- JobRunner now yields unsuccessful Finish when there is an
  unsuccessful job.
- Remove remaning StopIteration, since they are deprecated.
- Create file "constants" JOBS_FILE, LOG_URL.
- Remove ForwardModelStatus.dump as it was replaced by File reporter.

mortalisk committed:
Make a utility method to wait for side effects in tests

Conflicts:
	python/tests/res/job_queue/test_job_dispatch.py
	python/tests/utils/__init__.py

markusdregi commented on python/job_runner/cli.py:
```diff
@@ -0,0 +1,45 @@
+
+        if not os.path.exists(run_path):
+            sys.stderr.write("No such directory: %s\n" % run_path)
+            sys.exit(1)
```
Could not these two lines be replaced by:

```python
sys.exit("No such directory: {}".format(run_path))
```

markusdregi reviewed (approved):
Looks very good! :+1: I've written some issues to further extend this work. But I think all of those things should be addressed in new and smaller PR's...

Only thing is whether we should merge the `stderr` message and the system-exit call as commented below.
